### PR TITLE
Add monasca barclamp

### PR DESCRIPTION
**Pull request by samirjorina**
**Highlights:**
- all backend services are managed by Ansible-based monasca-installer (https://gitlab.com/monasca-installer).
- currently only single-node deployment is supported (clustered deployment is work in progress)
- monasca backend services should use dedicated 'monitoring' network

**Deployment:**
- monasca-master role should be deployed on the admin (AKA crowbar) node - Ansible uses SSH keys, which are available there
- monasa-server should be deployed on a dedicated, clean node
- monasca-metric-agent and monasca-log-agent can be deployed to any node(s)

### CODE REVIEW

File: crowbar_framework/app/models/monasca_service.rb [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    # Turn off multi proposal support till it really works and people ask for it.
+    def self.allow_multiple_proposals?
+      false
+    end
+
+    def role_constraints
+      {
+        "monasca-metric-agent" => {
+          "unique" => false,
+          "admin" => true,
```
Comment (dirkmueller) [inline crowbar_framework/app/models/monasca_service.rb]: should be false

File: crowbar_framework/app/models/monasca_service.rb [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+          "count" => -1,
+          "exclude_platform" => {
+            "suse" => "< 12.1",
+            "windows" => "/.*/"
+          }
+        },
+        "monasca-log-agent" => {
+          "unique" => false,
+          "admin" => true,
```
Comment (dirkmueller) [inline crowbar_framework/app/models/monasca_service.rb]: should be false

File: crowbar_framework/app/models/monasca_service.rb [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+          "count" => -1,
+          "exclude_platform" => {
+            "suse" => "< 12.1",
```
Comment (dirkmueller) [inline crowbar_framework/app/models/monasca_service.rb]: 12.2

File: crowbar_framework/app/models/monasca_service.rb [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+        "monasca-server" => {
+          "unique" => false,
+          # TODO: change for cluster
+          # "count" => -1,
+          # "cluster" => true,
+          "count" => 1,
+          "cluster" => false,
+          "admin" => false,
+          "exclude_platform" => {
+            "suse" => "< 12.1",
```
Comment (dirkmueller) [inline crowbar_framework/app/models/monasca_service.rb]: 12.2

File: crowbar_framework/app/models/monasca_service.rb [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+            "windows" => "/.*/"
+          }
+        },
+        "monasca-master" => {
+          "unique" => false,
+          "count" => 1,
+          "cluster" => false,
+          "admin" => true,
+          "exclude_platform" => {
+            "suse" => "< 12.1",
```
Comment (dirkmueller) [inline crowbar_framework/app/models/monasca_service.rb]: 12.2

File: crowbar_framework/app/models/monasca_service.rb [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+    nodes = NodeObject.all
+
+    server_nodes = nodes.select { |n| n.intended_role == "monitoring" }
+    server_nodes = [nodes.first] if server_nodes.empty?
+    server_nodes = server_nodes.map { |n| n.name }.compact
+
+    agent_nodes = nodes.map { |n| n.name }.compact
+
+    master_nodes = nodes.select { |n| n.intended_role == "admin" || n.name.start_with?("crowbar.") }
```
Comment (dirkmueller) [inline crowbar_framework/app/models/monasca_service.rb]: this doesn't look correct. nothing production critical should be deployed on the admin node.

File: monasca.yml [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+  requires:
+    - '@crowbar'
+    - 'database'
+    - 'keystone'
+  member:
+    - 'openstack'
+
+crowbar:
+  layout: 1
+  order: 98
```
Comment (dirkmueller) [inline monasca.yml]: that doesn't look correct, I would put it between keystone and glance..

File: chef/cookbooks/monasca/templates/default/cmm-hosts-single.erb [hunk 8]
```diff
@@ -0,0 +1,176 @@
+# Copyright 2017 FUJITSU LIMITED
```
Comment (dirkmueller) [inline chef/cookbooks/monasca/templates/default/cmm-hosts-single.erb]: the file should not have cmm in its name..

File: chef/cookbooks/monasca/templates/default/cmm-hosts-cluster.erb [hunk 9]
```diff
@@ -0,0 +1,286 @@
+# Copyright 2017 FUJITSU LIMITED
```
Comment (dirkmueller) [inline chef/cookbooks/monasca/templates/default/cmm-hosts-cluster.erb]: file should not have cmm in its name

File: chef/cookbooks/monasca/templates/default/api-config.conf.erb [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+# port = 3306
+# database = mon
+# query = ""
+
+[database]
+url = <%= @database_connection %>
+
+[keystone_authtoken]
+auth_type = password
+user_domain_id = default
```
Comment (dirkmueller) [inline chef/cookbooks/monasca/templates/default/api-config.conf.erb]: this should likely be 

```
project_domain_name = <%= @keystone_settings["admin_domain"]%>
user_domain_name = <%= @keystone_settings["admin_domain"] %>

```
instead of setting the _id as a name

File: chef/cookbooks/monasca/templates/default/api-config.conf.erb [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+project_domain_id = default
+auth_url = <%= @keystone_settings['admin_auth_url'] %>
```
Comment (dirkmueller) [inline chef/cookbooks/monasca/templates/default/api-config.conf.erb]: internal_auth_url

File: chef/cookbooks/monasca/templates/default/api-config.conf.erb [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+auth_uri = <%= @keystone_settings['public_auth_url'] %>
+auth_version = <%= @keystone_settings['api_version_for_middleware'] %>
+username = <%= @keystone_settings['service_user'] %>
+password = <%= @keystone_settings['service_password'] %>
+project_name = <%= @keystone_settings['service_tenant'] %>
```
Comment (dirkmueller) [inline chef/cookbooks/monasca/templates/default/api-config.conf.erb]: missing

```
region_name = <%= @keystone_settings['endpoint_region'] %>
```

File: chef/data_bags/crowbar/template-barbican.schema [hunk 13]
```diff
@@ -2,7 +2,7 @@
   "required": true,
   "type": "map",
   "mapping": {
-    "id": { "type": "str", "required": true, "pattern": "/^barbican-|^template-barbican$/" },
+    "id": { "type": "str", "required": true, "pattern": "/^mybarclamp-|^template-mybarclamp$/" },
```
Comment (dirkmueller) [inline chef/data_bags/crowbar/template-barbican.schema]: that change looks unrelated and wrong..

File: crowbar_framework/app/models/monasca_service.rb (hunk 1 above)
Comment (samirjorina) [inline crowbar_framework/app/models/monasca_service.rb]: done, see https://github.com/FujitsuEnablingSoftwareTechnologyGmbH/crowbar-openstack/pull/35

File: crowbar_framework/app/models/monasca_service.rb (hunk 6 above)
Comment (samirjorina) [inline crowbar_framework/app/models/monasca_service.rb]: monasca_master role is responsible for installing  and running the ansible installer, having this on the admin node allows ansible to reuse existing SSH key to connect to other nodes

File: monasca.yml (hunk 7 above)
Comment (samirjorina) [inline monasca.yml]: this was introduced in the barclamp skeleton, no idea why github shows that as a change file... I guess we'll need to resolve conflicts

File: chef/cookbooks/monasca/templates/default/cmm-hosts-single.erb (hunk 8 above)
Comment (samirjorina) [inline chef/cookbooks/monasca/templates/default/cmm-hosts-single.erb]: waiting for internal review: https://github.com/FujitsuEnablingSoftwareTechnologyGmbH/crowbar-openstack/pull/35

File: chef/cookbooks/monasca/templates/default/api-config.conf.erb (hunk 10 above)
Comment (samirjorina) [inline chef/cookbooks/monasca/templates/default/api-config.conf.erb]: recipe and config for api are not used, removing in https://github.com/FujitsuEnablingSoftwareTechnologyGmbH/crowbar-openstack/pull/35

File: chef/cookbooks/horizon/files/default/grafana-monasca.json [hunk 14]
```diff
@@ -0,0 +1,4751 @@
+{
+  "id": "sid-14393838984416185642089112666",
+  "title": "CMM Service",
```
Comment (dirkmueller) [inline chef/cookbooks/horizon/files/default/grafana-monasca.json]: The reference to CMM needs to be removed

File: chef/cookbooks/monasca/recipes/metric_agent.rb [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+#
+
+package "openstack-monasca-agent"
+
+agent_settings = node[:monasca][:metric_agent]
+agent_keystone = agent_settings[:keystone]
+
+keystone_settings = KeystoneHelper.keystone_settings(node, @cookbook_name)
+
+monasca_servers = search(:node, "roles:monasca-server") || []
```
Comment (Evalle) [inline chef/cookbooks/monasca/recipes/metric_agent.rb]: The weird thing is that monasca barclamp will apply "successfully" if I apply it the second time (right after this error will appear).

Comment (jgrassler) [inline chef/cookbooks/monasca/recipes/metric_agent.rb]: There may be a bit of an ordering issue here. Let me try and figure out a way to resolve it...

Comment (jgrassler) [inline chef/cookbooks/monasca/recipes/metric_agent.rb]: We discussed this one in our standup today, and there are two changes this needs:

1) Use `node_search_with_cache()` rather than `search` (that will work better on nodes that do not have the `monasca-server` role applied.

2) Add error handling for an empty list nonetheless. `node_search_with_cache()` may fail as well. In that case do the safe thing and just `return` after logging an an informational message indicating the problem and that this recipe will be processed again in the next run.

Comment (matrixik) [inline chef/cookbooks/monasca/recipes/metric_agent.rb]: Proposed fix: https://github.com/FujitsuEnablingSoftwareTechnologyGmbH/crowbar-openstack/pull/39

File: chef/cookbooks/horizon/files/default/grafana-monasca.json (hunk 14 above)
Comment (matrixik) [inline chef/cookbooks/horizon/files/default/grafana-monasca.json]: Will be fixed by https://github.com/FujitsuEnablingSoftwareTechnologyGmbH/crowbar-openstack/pull/37

File: chef/cookbooks/monasca/attributes/default.rb [hunk 16]
```diff
@@ -13,14 +14,41 @@
...
+# metric-agent default service settings
+default[:monasca][:metric_agent][:service_name] = monasca_metric_agent_service_name
+default[:monasca][:metric_agent][:user] = "monasca-agent"
+default[:monasca][:metric_agent][:group] = "monasca"
+default[:monasca][:metric_agent][:debug] = monasca_debug
+default[:monasca][:metric_agent][:agent_service_name] = "openstack-monasca-agent"
+
+# log-agent default service settings
+default[:monasca][:log_agent][:service_name] = monasca_log_agent_service_name
+default[:monasca][:log_agent][:user] = "monasca-log-agent"
```
Comment (jgrassler) [inline chef/cookbooks/monasca/attributes/default.rb]: With https://build.opensuse.org/request/show/486330 merged we no longer need a user/group for monasca-log-agent. In fact trying to create files owned by this non-existent user/group breaks deployment, so this needs to go.

File: chef/cookbooks/monasca/recipes/log_agent.rb [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+end
+
+monasca_log_api_url = MonascaHelper.log_api_public_url(monasca_server)
+
+log_agent_dimensions = {
+  hostname: node["hostname"]
+}
+
+directory "/var/log/monasca-log-agent/" do
+  owner log_agent_settings["user"]
```
Comment (jgrassler) [inline chef/cookbooks/monasca/recipes/log_agent.rb]: With https://build.opensuse.org/request/show/486330 merged and monasca-log-agent running as root we no longer need to make this files owned by a dedicated user. In fact trying to create files owned by this non-existent user breaks deployment, so this needs to go (make them owned by `root:root` instead).

File: chef/cookbooks/monasca/recipes/log_agent.rb [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+  group log_agent_settings["group"]
+  mode 0o755
+  recursive true
+end
+directory "/etc/monasca-log-agent/" do
+  owner log_agent_settings["user"]
```
Comment (jgrassler) [inline chef/cookbooks/monasca/recipes/log_agent.rb]: This needs to be owned by root:root (see above).

File: chef/cookbooks/monasca/recipes/log_agent.rb [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+    log_dirs =
+      Dir.entries("/var/log")
+         .select { |e| File.directory?("/var/log/#{e}") && !(e == "." || e == "..") }
+    log_dirs.each { |d| log_files["/var/log/#{d}/**/*.log"] = d.downcase }
+  end
+end
+
+template "/etc/monasca-log-agent/agent.conf" do
+  source "log-agent.conf.erb"
+  owner log_agent_settings["user"]
```
Comment (jgrassler) [inline chef/cookbooks/monasca/recipes/log_agent.rb]: This needs to be owned by root:root (see above).

File: chef/cookbooks/monasca/recipes/log_agent.rb (hunk 17 above)
Comment (matrixik) [inline chef/cookbooks/monasca/recipes/log_agent.rb]: https://github.com/FujitsuEnablingSoftwareTechnologyGmbH/crowbar-openstack/pull/42

**Comment by toabctl** [comment]
Is this still needed? If so it needs to be rebased. Otherwise, please close the PR

**Comment by Evalle** [comment]
Monasca barclamp now can be installed without this PR, so we already have all these changes in our code.

**Comment by dirkmueller** [comment]
continued in https://github.com/crowbar/crowbar-openstack/pull/919
